core/src/main/java/com/linecorp/armeria/internal/common/ByteBufBytes.java

```diff
@@ -0,0 +1,264 @@
+    public InputStream toInputStream() {
+        if (array != null) {
+            return new FastByteArrayInputStream(array);
+        } else {
+            return new ByteBufInputStream(buf.duplicate(), false);
+        }
+    }
+
+    @Override
+    public boolean isPooled() {
+        return pooled;
+    }
+
+    @Override
+    public ByteBuf byteBuf(ByteBufAccessMode mode) {
+        switch (requireNonNull(mode, "mode")) {
+            case DUPLICATE:
+                return buf.duplicate();
+            case RETAINED_DUPLICATE:
+                return buf.retainedDuplicate();
+            case FOR_IO:
+                if (buf.isDirect()) {
+                    return buf.retainedDuplicate();
+                }
+
+                final ByteBuf copy = newDirectByteBuf();
+                copy.writeBytes(buf, buf.readerIndex(), buf.readableBytes());
+                return copy;
+        }
+
+        throw new Error(); // Never reaches here.
+    }
+
+    @Override
+    public ByteBuf byteBuf(int offset, int length, ByteBufAccessMode mode) {
+        final int startIndex = buf.readerIndex() + offset;
+        switch (requireNonNull(mode, "mode")) {
+            case DUPLICATE:
+                return buf.slice(startIndex, length);
+            case RETAINED_DUPLICATE:
+                return buf.retainedSlice(startIndex, length);
+            case FOR_IO:
+                if (buf.isDirect()) {
+                    return buf.retainedSlice(startIndex, length);
+                }
+
+                final ByteBuf copy = newDirectByteBuf(length);
+                copy.writeBytes(buf, startIndex, length);
+                return copy;
+        }
+
+        throw new Error(); // Never reaches here.
+    }
+
+    private ByteBuf newDirectByteBuf() {
+        return newDirectByteBuf(buf.readableBytes());
+    }
+
+    private static ByteBuf newDirectByteBuf(int length) {
+        return PooledByteBufAllocator.DEFAULT.directBuffer(length);
+    }
+
+    @Override
+    public void touch(@Nullable Object hint) {
+        if (isPooled()) {
+            buf.touch(firstNonNull(hint, this));
+        }
+    }
+
+    @Override
+    public void close() {
+        // This is not thread safe, but an attempt to close one instance from multiple threads would fail
+        // with an IllegalReferenceCountException anyway.
+        if (!closed) {
+            closed = true;
+            if (pooled) {
+                buf.release();
+            }
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        // Calculate the hash code from the first 32 bytes.
+        int hash = 0;
+        final int bufStart = buf.readerIndex();
+        final int bufEnd = bufStart + Math.min(32, buf.readableBytes());
+        for (int i = bufStart; i < bufEnd; i++) {
+            hash = hash * 31 + buf.getByte(i);
+        }
+        return hash;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (!(obj instanceof ByteBufBytes)) {
```

Comment:
Previously, the content of pooled `HttpData` and array `HttpData` can be comparable.
But this only allows `ByteBufBytes`. Could we make `ByteBufBytes` comparable to `ByteArrayBytes`? It seems inefficient but useful for testing.

Reply:
That's a good idea. Let's use `Bytes` then. 😄 
Also, I changed `equals` in `ByteBufHttpData` and `ByteBufArrayData` to allow comparison to each other.